rescan_test.go

```diff
@@ -504,3 +504,188 @@ func (ctx *rescanTestContext) waitUntilSynced() {
+	// Note the block which we'll revert back to.
+	bestBlock, err := ctx.chain.BestBlock()
+	if err != nil {
+		ctx.t.Fatalf("unable to retrieve best block: %v", err)
+	}
+
+	// Modify the mocked chain such that it fails to retrieve block filters.
+	ctx.chain.setFailGetFilter(true)
+
+	// Adding multiple new blocks to the chain should result in the first
+	// retry block being queried until successful.
+	block1 := ctx.chain.addNewBlock(true)
+	ctx.assertFilterQueried(block1.Hash)
+	block2 := ctx.chain.addNewBlock(true)
+	ctx.assertFilterQueried(block1.Hash)
+
+	// We'll then reorg back to a height where the new blocks we've added
+	// and retried are stale, causing us to no longer retry them. This won't
+	// deliver any disconnected notifications since the blocks were never
+	// connected.
+	ctx.chain.rollbackToHeight(bestBlock.Height, true)
+
+	// Revert the mocked chain so that block filters can be retrieved
+	// successfully.
+	ctx.chain.setFailGetFilter(false)
+
+	// We'll then add two new blocks from a different chain to ensure the
+	// rescan can properly follow it.
+	newBlock1 := ctx.chain.addNewBlockWithHeader(
+		&wire.BlockHeader{PrevBlock: bestBlock.Hash}, true,
+	)
+	if newBlock1.Hash == block1.Hash {
+		ctx.t.Fatal("expected different hashes for blocks on " +
+			"different chains")
+	}
+	ctx.assertFilterQueried(newBlock1.Hash)
+	ctx.recvBlockConnected(newBlock1)
+
+	newBlock2 := ctx.chain.addNewBlock(true)
+	if newBlock2.Hash == block2.Hash {
+		ctx.t.Fatal("expected different hashes for blocks on " +
+			"different chains")
+	}
+	ctx.assertFilterQueried(newBlock2.Hash)
+	ctx.recvBlockConnected(newBlock2)
+}
+
+// TestRescanRetryBlocksAfterCatchingUp ensures the rescan stops retrying blocks
+// after catching up with the chain. Catching up assumes that we already have an
+// accurate representation of the chain.
+func TestRescanRetryBlocksAfterCatchingUp(t *testing.T) {
+	t.Parallel()
+
+	// Dummy options to allow the Rescan to fetch block filters.
+	ctx := newRescanTestContext(t, 10, []RescanOption{
+		StartTime(time.Time{}),
+		WatchInputs(InputWithScript{}),
+	}...)
+	ctx.start()
+	defer ctx.stop()
+
+	ctx.waitUntilSynced()
+
+	// Modify the mocked chain such that it fails to retrieve block filters.
+	ctx.chain.setFailGetFilter(true)
+
+	// Adding multiple new blocks to the chain should result in the first
+	// retry block being queried until successful.
+	block1 := ctx.chain.addNewBlock(true)
+	ctx.assertFilterQueried(block1.Hash)
+
+	// We'll prevent notifying new blocks and instead notify an invalid
+	// block which should prompt the rescan to catch up with the chain
+	// manually.
+	block2 := ctx.chain.addNewBlock(false)
+	block3 := ctx.chain.addNewBlock(false)
+	ctx.chain.ntfnChan <- blockntfns.NewBlockConnected(wire.BlockHeader{}, 0)
+
+	// Revert the mocked chain so that block filters can be retrieved
+	// successfully.
+	ctx.chain.setFailGetFilter(false)
+
+	// The notifications for all the blocks added above should now be
```

Comment:
should we assert that the rescan loop is also fetching the blocks from the chain source here?

Reply:
Doesn't seem like a useful assertion to make at the moment?

Reply:
We assert that filters are queried, so could also assert that blocks are, but no blocker.